libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp

```diff
@@ -183,16 +183,29 @@ class flystel_prime_field_gadget : public gadget<libff::Fr<ppT>>
     void generate_r1cs_witness();
 };
 
-// get the MDS matrix for each allowed dimension: 2,3 or 4
-template<typename ppT>
-std::array<std::array<libff::Fr<ppT>, 2>, 2> anemoi_permutation_mds_2x2(
-    const libff::Fr<ppT> g);
-template<typename ppT>
-std::array<std::array<libff::Fr<ppT>, 3>, 3> anemoi_permutation_mds_3x3(
-    const libff::Fr<ppT> g);
-template<typename ppT>
-std::array<std::array<libff::Fr<ppT>, 4>, 4> anemoi_permutation_mds_4x4(
-    const libff::Fr<ppT> g);
+// MDS matrix for each allowed dimension: 2,3 or 4
+template<typename ppT, size_t NumStateColumns_L> class anemoi_permutation_mds;
+
+template<typename ppT> class anemoi_permutation_mds<ppT, 2>
+{
+public:
+    static std::array<std::array<libff::Fr<ppT>, 2>, 2> permutation_mds(
+        const libff::Fr<ppT> g);
```

Comment:
ATM I am only using `anemoi_parameters<ppT>::multiplicative_generator_g`. I really like the idea of using the latter implicitly instead of explicitly passing an extra input parameter `g` and I am tempted to apply the change (in fact, I already did but haven't pushed it yet). Note though that for some unit tests I am using the `debug_parameters_bls12_381` class which incidentally has the same value for `g=7`, but which can potentially be changed. So I wonder if I shouldn't keep the input parameter `g` after all, in order to accommodate a potential change of its value e.g. for unit testing. Happy to go both ways, so interested to hear your opinion.

Reply:
Ah yes, that's a good point.  The aemoi_parameters is now a template parameter.  I'd forgotten that.  I guess we could make it a parameter of this class too, but it might get complex.  Ideally we'd extract the ppT part from the anemoi_parameters as well, so there were only 2 parameters.

Feel free to just put a note in the comments instead of making the change.  As you prefer.  This is all "internal" to the anemoi code anyway so it's easy to go back and change later without affecting too much other code.

Reply:
Ok, kept the input parameter `g` but added a TODO node as suggested.